### PR TITLE
[Topi] Fast mode in take op

### PR DESCRIPTION
**Pull request by hlu1**
RFC https://github.com/dmlc/tvm/issues/3303
@antinucleon, @icemelon9, @ajtulloch please review.

### CODE REVIEW

File: topi/python/topi/transform.py [hunk 1]
```diff
@@ -264,6 +264,7 @@ def take(a, indices, axis=None, mode="clip"):
     mode : str, optional
         Specifies how out-of-bound indices will behave.
         clip - clip to the range (default)
+        fast - no lip or wrap aroud
```
Comment (icemelon9) [inline topi/python/topi/transform.py]:
```suggestion
        fast - no clip or wrap aroud
```

File: topi/include/topi/transform.h [hunk 2]
```diff
@@ -641,6 +641,13 @@ inline Tensor take(const Tensor& a,
           auto idx = tvm::min(tvm::max(0, indices(out_index)), a_size - 1);
           return a(UnravelIndex(idx, a_shape));
         }, name, tag);
+  } else if (mode == "fast") {
+    LOG(WARNING) << "Fast mode segfaults when there are out-of-bounds indices. "
+                "Make sure input indices are in bound";
```
Comment (icemelon9) [inline topi/include/topi/transform.h]: fix indent

File: topi/python/topi/transform.py (hunk 1 above)
Comment (icemelon9) [reply topi/python/topi/transform.py]: Could you put this line below wrap?

File: include/tvm/relay/attrs/transform.h [hunk 3]
```diff
@@ -101,6 +101,7 @@ struct TakeAttrs : public tvm::AttrsNode<TakeAttrs> {
     TVM_ATTR_FIELD(mode).set_default("clip")
         .describe("Specify how out-of-bound indices will behave."
                   "clip - clip to the range (default)"
+                  "fast - no lip or wrap aroud"
```
Comment (icemelon9) [inline include/tvm/relay/attrs/transform.h]: Put this line below wrap.

**Comment by hlu1** [comment]
Typo and indent are fixed. Added doc re `relay.take`. Test cases added as well.

File: topi/python/topi/transform.py [hunk 4]
```diff
@@ -265,6 +265,7 @@ def take(a, indices, axis=None, mode="clip"):
         Specifies how out-of-bound indices will behave.
         clip - clip to the range (default)
         wrap - wrap around the indices
+        fast - no clip or wrap aroud
```
Comment (antinucleon) [inline topi/python/topi/transform.py]: around (boundary should be check by user)

**Comment by icemelon9** [comment]
Could you add test cases to `tests/python/relay/test_op_level3.py`?

File: topi/python/topi/transform.py (hunk 4 above)
Comment (hlu1) [reply topi/python/topi/transform.py]: Added

**Comment by hlu1** [comment]
> Could you add test cases to `tests/python/relay/test_op_level3.py`?

Added.

File: include/tvm/relay/attrs/transform.h [hunk 5]
```diff
@@ -101,7 +101,8 @@ struct TakeAttrs : public tvm::AttrsNode<TakeAttrs> {
     TVM_ATTR_FIELD(mode).set_default("clip")
         .describe("Specify how out-of-bound indices will behave."
                   "clip - clip to the range (default)"
-                  "wrap - wrap around the indices");
+                  "wrap - wrap around the indices"
+                  "fast - no clip or wrap aroud (user must make sure indices are in-bound)");
```
Comment (icemelon9) [inline include/tvm/relay/attrs/transform.h]:
```suggestion
                  "fast - no clip or wrap around (user must make sure indices are in-bound)");
```

File: topi/python/topi/transform.py [hunk 6]
```diff
@@ -265,6 +265,7 @@ def take(a, indices, axis=None, mode="clip"):
         Specifies how out-of-bound indices will behave.
         clip - clip to the range (default)
         wrap - wrap around the indices
+        fast - no clip or wrap aroud (user must make sure indices are in-bound)
```
Comment (icemelon9) [inline topi/python/topi/transform.py]:
```suggestion
        fast - no clip or wrap around (user must make sure indices are in-bound)
```

File: include/tvm/relay/attrs/transform.h (hunk 5 above)
Comment (hlu1) [reply include/tvm/relay/attrs/transform.h]: Oops. Fixed.

File: topi/python/topi/transform.py (hunk 6 above)
Comment (hlu1) [reply topi/python/topi/transform.py]: Fixed

File: python/tvm/relay/op/transform.py [hunk 7]
```diff
@@ -221,6 +221,7 @@ def take(data, indices, axis=None, mode="clip"):
         Specifies how out-of-bound indices will behave [clip, wrap].
```
Comment (icemelon9) [inline python/tvm/relay/op/transform.py]:
```suggestion
        Specifies how out-of-bound indices will behave [clip, wrap, fast].
```

Comment (hlu1) [reply python/tvm/relay/op/transform.py]: Added

**Comment by icemelon9** [comment]
Thanks @hlu1 @antinucleon . this is now merged.

**Pull request merged**